lib/pbench/agent/tool_group.py

```diff
@@ -0,0 +1,158 @@
+
+          "toolnames" - each tool name is the key, with separate dictionaries
+          for each registered host
+
+          "hostnames" - each registered host is the key, with separate
+          dictionaries for each tool registered on that host
+
+          "labels" - each registered host name, that has a label, is the key,
+          and the label as the value; if a host is not labeled, it does not
+          show up in this dictionary
+
+        Raises BadToolGroup via the verify_tool_group() method on error.
+        """
+        self.tg_dir = self.verify_tool_group(group)
+        self.group = group
+
+        # __trigger__
+        try:
+            _trigger = (self.tg_dir / "__trigger__").read_text()
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                raise
+            # Ignore missing trigger file
+            self.trigger = None
+        else:
+            if len(_trigger) == 0:
+                # Ignore empty trigger file contents
+                self.trigger = None
+            else:
+                self.trigger = _trigger
+
+        # toolnames - Dict with tool name as the key, dictionary with host
+        # names and parameters for each host
+        self.toolnames = {}
+        # hostnames - Dict with host name as the key, dictionary with tool
+        # names and parameters for each tool
+        self.hostnames = {}
+        self.labels = {}
+        for hdirent in os.listdir(self.tg_dir):
+            if hdirent == "__trigger__":
+                # Ignore handled above
+                continue
+            if not (self.tg_dir / hdirent).is_dir():
+                # Ignore wayward non-directory files
+                continue
+            # We assume this directory is a hostname.
+            host = hdirent
+            if host not in self.hostnames:
+                self.hostnames[host] = {}
+            for tdirent in os.listdir(self.tg_dir / host):
+                if tdirent == "__label__":
+                    self.labels[host] = (
+                        (self.tg_dir / host / tdirent).read_text().strip()
+                    )
+                    continue
+                if tdirent.endswith("__noinstall__"):
+                    # FIXME: ignore "noinstall" for now, tools are going to be
+                    # in containers so this does not make sense going forward.
+                    continue
+                # This directory entry is the name of a tool.
+                tool = tdirent
+                tool_opts_raw_lines = (
+                    (self.tg_dir / host / tool).read_text().split("\n")
+                )
+                tool_opts_lines = []
+                for line_raw in tool_opts_raw_lines:
+                    line = line_raw.strip()
+                    if not line:
+                        # Ignore blank lines
+                        continue
+                    tool_opts_lines.append(line)
+                tool_opts = " ".join(tool_opts_lines)
```

Comment:
I think you can do this in two code lines:  read the text, and do a regex substitution which replaces the newlines with blanks while removing blank lines, e.g.:
```
                tool_opts_text = (self.tg_dir / host / tool).read_text().strip()
                tool_opts = re.sub(r'\n\s*', ' ', tool_opts_text)
```